finality-aleph/src/base_protocol/handler.rs

```diff
@@ -45,18 +102,98 @@ impl Handler {
             })
             .collect()
     }
-}
 
-#[cfg(test)]
-mod tests {
-    use crate::{aleph_primitives::Block, base_protocol::handler::Handler};
-
-    #[test]
-    fn initially_no_peers() {
-        let handler = Handler::new();
-        assert!(
-            handler.peers_info::<Block>().is_empty(),
-            "there should be no peers initially"
-        );
+    fn verify_connection(
+        &self,
+        peer_id: PeerId,
+        handshake: Vec<u8>,
+        is_inbound: bool,
+    ) -> Result<Roles, ConnectError> {
+        let handshake = BlockAnnouncesHandshake::<B>::decode_all(&mut &handshake[..])
+            .map_err(|_| ConnectError::BadlyEncodedHandshake)?;
+        if handshake.genesis_hash != self.genesis_hash {
+            return Err(ConnectError::BadHandshakeGenesis);
+        }
+
+        if self.peers.contains_key(&peer_id) {
+            return Err(ConnectError::PeerAlreadyConnected);
+        }
+
+        if !self.reserved_nodes.contains(&peer_id) {
+            // check slot constraints for full non-reserved nodes
+            if is_inbound
+                && handshake.roles.is_full()
+                && self.num_full_in_peers >= self.max_full_in_peers
+            {
+                return Err(ConnectError::TooManyFullInboundPeers);
+            }
+            if !is_inbound
+                && handshake.roles.is_full()
+                && self.num_full_out_peers >= self.max_full_out_peers
+            {
+                return Err(ConnectError::TooManyFullOutboundPeers);
+            }
+            // check slot constraints for light nodes
+            if handshake.roles.is_light() && self.num_light_peers >= self.max_light_peers {
+                return Err(ConnectError::TooManyLightPeers);
+            }
+        }
+
+        Ok(handshake.roles)
+    }
+
+    pub fn on_peer_connect(
+        &mut self,
+        peer_id: PeerId,
+        handshake: Vec<u8>,
+        is_inbound: bool,
+    ) -> Result<(), ConnectError> {
+        let role = self.verify_connection(peer_id, handshake, is_inbound)?;
+
+        // update peer sets
+        self.peers.insert(peer_id, PeerInfo { role, is_inbound });
+
+        if !self.reserved_nodes.contains(&peer_id) {
+            // update slots for full nodes
+            if is_inbound && role.is_full() {
+                self.num_full_in_peers += 1;
+            }
+            if !is_inbound && role.is_full() {
+                self.num_full_out_peers += 1;
+            }
+            // update slots of light nodes
+            if role.is_light() {
+                self.num_light_peers += 1;
+            }
+        }
+
+        // TODO(A0-3901): disseminate appropriate `SyncEvent`
```

Comment:
This is IO, so it should rather happen in `Service`. And the appropriate sinks are already there. This means this method should probably return the `PeerId` on success.

Same below.

Reply:
Right. Maybe just return `()` cuz service already has`peer_id` (and it implements `Copy` so no need to `peer_id.clone()`).

Reply:
I prefer returning, as then the calls look more natural imo, but this is fine as well.

Reply:
removed the `TODO`